heapster-saw/src/Verifier/SAW/Heapster/RustTypes.hs

```diff
@@ -812,12 +815,37 @@ mbSeparatePrx _ = mbSeparate
 mbAssoc :: prx1 ctx1 -> RAssign prx2 ctx2 -> RAssign prx3 ctx3 ->
            Mb (ctx1 :++: (ctx2 :++: ctx3)) a ->
            Mb ((ctx1 :++: ctx2) :++: ctx3) a
+mbAssoc ctx1 ctx2 ctx3 mb_a | Refl <- RL.appendAssoc ctx1 ctx2 ctx3 = mb_a
+{-
 mbAssoc ctx1 ctx2 ctx3 mb_a =
   mbCombine (RL.mapRAssign (const Proxy) ctx3) $
   mbCombine (RL.mapRAssign (const Proxy) ctx2) $
   fmap (mbSeparatePrx ctx2 ctx3) $
   mbSeparatePrx ctx1 (RL.append (RL.map (const Proxy) ctx2)
                       (RL.map (const Proxy) ctx3)) mb_a
+-}
+
+mbUnAssoc :: prx1 ctx1 -> RAssign prx2 ctx2 -> RAssign prx3 ctx3 ->
+             Mb ((ctx1 :++: ctx2) :++: ctx3) a ->
+             Mb (ctx1 :++: (ctx2 :++: ctx3)) a
+mbUnAssoc ctx1 ctx2 ctx3 mb_a | Refl <- RL.appendAssoc ctx1 ctx2 ctx3 = mb_a
+
+appendAssoc4 :: RAssign Proxy ctx1 -> RAssign Proxy ctx2 ->
+                RAssign Proxy ctx3 -> RAssign Proxy ctx4 ->
+                ctx1 :++: ((ctx2 :++: ctx3) :++: ctx4) :~:
+                ((ctx1 :++: ctx2) :++: ctx3) :++: ctx4
+appendAssoc4 ctx1 ctx2 ctx3 ctx4
+  | Refl <- RL.appendAssoc ctx1 (RL.append ctx2 ctx3) ctx4
+  , Refl <- RL.appendAssoc ctx1 ctx2 ctx3
+  = Refl
+
+
+mbAssoc4 :: RAssign Proxy ctx1 -> RAssign Proxy ctx2 ->
+            RAssign Proxy ctx3 -> RAssign Proxy ctx4 ->
+            Mb (ctx1 :++: ((ctx2 :++: ctx3) :++: ctx4)) a ->
+            Mb (((ctx1 :++: ctx2) :++: ctx3) :++: ctx4) a
+mbAssoc4 ctx1 ctx2 ctx3 ctx4 mb_a
+  | Refl <- appendAssoc4 ctx1 ctx2 ctx3 ctx4 = mb_a
```

Comment:
Should we leave a comment here saying these should be moved to Hobbits? If so, maybe it would be nice to put them at the top of the file. (Same with `mbAssoc` above and `mbCombineAssoc4`, `assocAppend4` below.)